bittide-instances/tests/Tests/ClockControlWb.hs

```diff
@@ -0,0 +1,226 @@
+    }
+
+dut ::
+  Circuit () (Df System (BitVector 8), CSignal System (ClockControlData LinkCount))
+dut =
+  withClockResetEnable
+    clockGen
+    resetGen
+    enableGen
+    $ circuit
+    $ \_unit -> do
+      (uartRx, jtag) <- idleSource -< ()
+      [uartBus, ccWb, dbgWb] <- processingElement peConfig -< jtag
+      (uartTx, _uartStatus) <- uartInterfaceWb d2 d2 uartSim -< (uartBus, uartRx)
+      [ccd0, ccd1] <-
+        csDupe
+          <| clockControlWb
+            margin
+            framesize
+            (pure linkMask)
+            (pure <$> dataCounts)
+          -< ccWb
+      cm <- cSigMap clockMod -< ccd0
+      _dbg <- debugRegisterWb (pure debugRegisterConfig) -< (dbgWb, cm)
+      idC -< (uartTx, ccd1)
+ where
+  (iMem, dMem) =
+    $( do
+        root <- runIO $ findParentContaining "cabal.project"
+        let
+          elfDir = root </> firmwareBinariesDir "riscv32imc-unknown-none-elf" Release
+          elfPath = elfDir </> "clock-control-wb"
+          iSize = 8 * 1024 -- 16 KB
+          dSize = 64 * 1024 -- 256 KB
+        memBlobsFromElf BigEndian (Just iSize, Just dSize) elfPath Nothing
+     )
+
+  peConfig =
+    PeConfig
+      (0b100 :> 0b010 :> 0b001 :> 0b110 :> 0b111 :> Nil)
+      (Reloadable $ Blob iMem)
+      (Reloadable $ Blob dMem)
+
+-- | Parse the output of the UART
+resultParser :: Parser SerialResult
+resultParser = do
+  linkCountResult <- expectField "nLinks"
+  linkMaskResult <- expectField "linkMask"
+  linkMaskPopcntResult <- expectField "linkMaskPopcnt"
+  reframingEnabledResult <- expectField "reframingEnabled"
+  linksStableResult <- expectField "linksStable"
+  linksSettledResult <- expectField "linksSettled"
+  dataCountsResult <- expectField "dataCounts"
+  clockModResult <- expectField "clockMod"
+  return
+    SerialResult
+      { linkCount = linkCountResult
+      , linkMask = linkMaskResult
+      , linkMaskPopcnt = linkMaskPopcntResult
+      , reframingEnabled = reframingEnabledResult
+      , linksStable = linksStableResult
+      , linksSettled = linksSettledResult
+      , dataCounts = dataCountsResult
+      , clockMod = clockModResult
+      }
+
+{- | A parser that parses a string in the form "name: value\n"
+The parsed name should be equal to the given fieldName
+-}
+expectField :: (Read a) => String -> Parser a
+expectField fieldName = do
+  (actualName, value) <- fieldParser
+  if actualName == fieldName
+    then pure value
+    else do
+      parserFail
+        [__i|
+        expectField failed to match the expected field name.
+        Expected field: #{fieldName}
+        Actual field:   #{actualName}|]
+
+-- | Parse a field of the form "name: value\n"
+fieldParser :: (Read a) => Parser (String, a)
+fieldParser = do
+  name <- manyTill anyChar (try (string ": "))
+  value <- manyTill anyChar (try (string "\n"))
+  return (name, read value)
+
+{- | Return the beginning of a list until you detect a certain sublist
+That substring is not included in the result.
+-}
```

Comment:
Comment is now wrong

Reply:
It returns the list before the sublist and after the sublist such that `input == before <> sublist <> after`